flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointFailureHandlingITCase.java

```diff
@@ -0,0 +1,311 @@
+
+    @Rule
+    public final MiniClusterWithClientResource miniClusterResource =
+            new MiniClusterWithClientResource(
+                    new MiniClusterResourceConfiguration.Builder()
+                            .setNumberTaskManagers(PARALLELISM)
+                            .setNumberSlotsPerTaskManager(1)
+                            .build());
+
+    @Test
+    public void testCheckpointSuccessAfterFailure() throws Exception {
+        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+        TestCheckpointStorage storage =
+                new TestCheckpointStorage(
+                        new JobManagerCheckpointStorage(), sharedObjects, temporaryFolder);
+
+        configure(env, storage);
+        buildGraph(env);
+
+        JobClient jobClient = env.executeAsync();
+        JobID jobID = jobClient.getJobID();
+        MiniCluster miniCluster = miniClusterResource.getMiniCluster();
+
+        waitForJobStatus(jobClient, singletonList(RUNNING), fromNow(Duration.ofSeconds(30)));
+        waitForAllTaskRunning(miniCluster, jobID, false);
+
+        triggerFailingCheckpoint(jobID, TestException.class, miniCluster);
+
+        miniCluster.triggerCheckpoint(jobID).get();
+    }
+
+    private void configure(StreamExecutionEnvironment env, TestCheckpointStorage storage) {
+        // enable checkpointing but only via API
+        env.enableCheckpointing(Long.MAX_VALUE, CheckpointingMode.EXACTLY_ONCE);
+
+        env.getCheckpointConfig().setCheckpointStorage(storage);
+
+        // use non-snapshotting backend to test channel state persistence integration with
+        // checkpoint storage
+        env.setStateBackend(new MockStateBackend(true));
+
+        env.getCheckpointConfig().enableUnalignedCheckpoints();
+
+        env.getCheckpointConfig().setAlignedCheckpointTimeout(Duration.ZERO); // speed-up
+
+        // failures are emitted by the storage
+        env.getCheckpointConfig().setTolerableCheckpointFailureNumber(Integer.MAX_VALUE);
+
+        // DoP > 1 is required for some barriers to lag
+        env.setParallelism(PARALLELISM);
+
+        // no chaining to have input channels (doesn't matter local or remote)
+        env.disableOperatorChaining();
+    }
+
+    private void buildGraph(StreamExecutionEnvironment env) {
+        // with zero alignment timeout some steps here are not strictly necessary currently, but
+        // this may change in the future
+        env.fromSource(
+                        new NumberSequenceSource(0, Long.MAX_VALUE),
+                        WatermarkStrategy.noWatermarks(),
+                        "num-source")
+                // source is not parallel, so keyBy to send to all down-streams
+                .keyBy(value -> value)
+                // exert back-pressure
+                .map(
+                        value -> {
+                            Thread.sleep(0, 100);
```

Comment:
nit: It is still Thread.sleep(1), look into its implementation. TBH, I don't know what's the point of the method in jvm.

Reply:
I think it's platform-specific; I don't mind changing to 1ms though.

Reply:
It does not really matter for this case, but I don't think it is platform-specific. It is not a native method. It does have an implementation that always converts the nanos part into milliseconds. 

It is true that the actual time how long a thread sleeps is platform-specific and depends on system timers and interruptions. From a brief investigation I did some time ago on Linux systems the minimal time of an "actual" time a thread can sleep is ~1ms whereas on windows ~10-15ms.

Reply:
Yes, by platform I meant JRE + OS + hardware (and when I briefly checked open-j9 this method was native and linux had suitable syscalls).
But anyways I've already changed the code to 1ms.

Reply:
`Thread.sleep(long millis)` is native, `Thread.sleep(long millis, long nanos)` is not ;): https://github.com/openjdk/jdk/blob/b6a729a79b2ca65ec3dc26af1ee8dcd1cd3afa41/jdk/src/java.base/share/classes/java/lang/Thread.java#L327 (even on current master)

Reply:
This is open-jdk, I was referring to open-j9: https://github.com/eclipse-openj9/openj9/blob/master/jcl/src/java.base/share/classes/java/lang/Thread.java#L1141

Reply:
ah, ok... Thanks, good to know!